File: backend/modules/lesson_catalog/public.py
"""
Lesson Catalog Module - Public Interface

Protocol definition and dependency injection provider.
This is the only interface other modules should import from.
"""

from typing import Protocol

from modules.content.public import ContentProvider

from .service import (
    BrowseLessonsResponse,
    CatalogStatistics,
    LessonCatalogService,
    LessonDetail,
    LessonSummary,
    RefreshCatalogResponse,
    SearchLessonsResponse,
)


class LessonCatalogProvider(Protocol):
    """Protocol defining the lesson catalog module's public interface."""

    def browse_lessons(self, user_level: str | None = None, limit: int = 100) -> BrowseLessonsResponse: ...
    def get_lesson_details(self, lesson_id: str) -> LessonDetail | None: ...
    def search_lessons(
        self,
        query: str | None = None,
        user_level: str | None = None,
        min_duration: int | None = None,
        max_duration: int | None = None,
        ready_only: bool = False,
        limit: int = 100,
        offset: int = 0,
    ) -> SearchLessonsResponse: ...
    def get_popular_lessons(self, limit: int = 10) -> list[LessonSummary]: ...
    def get_catalog_statistics(self) -> CatalogStatistics: ...
    def refresh_catalog(self) -> RefreshCatalogResponse: ...


def lesson_catalog_provider(content: ContentProvider) -> LessonCatalogProvider:
    """
    Dependency injection provider for lesson catalog services.

    Args:
        content: Content service instance (built with same session as caller).

    Returns:
        LessonCatalogService instance that implements the LessonCatalogProvider protocol.
    """
    return LessonCatalogService(content)


__all__ = [
    "BrowseLessonsResponse",
    "CatalogStatistics",
    "LessonCatalogProvider",
    "LessonDetail",
    "LessonSummary",
    "RefreshCatalogResponse",
    "SearchLessonsResponse",
    "lesson_catalog_provider",
]

File: mobile/modules/lesson_catalog/public.ts
/**
 * Lesson Catalog Public Interface
 *
 * The only interface other modules should import from.
 * Pure forwarder - no logic, just selects/forwards service methods.
 */

import { LessonCatalogService } from './service';
import { LessonCatalogRepo } from './repo';
import type {
  LessonSummary,
  LessonDetail,
  BrowseLessonsResponse,
  LessonFilters,
  CatalogStatistics,
  PaginationInfo,
} from './models';

// Public interface protocol
export interface LessonCatalogProvider {
  // Only expose what learning_session actually needs
  getLessonDetail(lessonId: string): Promise<LessonDetail | null>;

  // Internal methods for lesson_catalog's own screens (not cross-module)
  browseLessons(
    filters?: LessonFilters,
    pagination?: Omit<PaginationInfo, 'hasMore'>
  ): Promise<BrowseLessonsResponse>;
  searchLessons(
    query: string,
    filters?: LessonFilters,
    pagination?: Omit<PaginationInfo, 'hasMore'>
  ): Promise<BrowseLessonsResponse>;
  getPopularLessons(limit?: number): Promise<LessonSummary[]>;
  getCatalogStatistics(): Promise<CatalogStatistics>;
  refreshCatalog(): Promise<{
    refreshedLessons: number;
    totalLessons: number;
    timestamp: string;
  }>;
  checkHealth(): Promise<boolean>;
}

// Service instance (singleton)
let serviceInstance: LessonCatalogService | null = null;

function getServiceInstance(): LessonCatalogService {
  if (!serviceInstance) {
    const repo = new LessonCatalogRepo();
    serviceInstance = new LessonCatalogService(repo);
  }
  return serviceInstance;
}

// Public provider function
export function lessonCatalogProvider(): LessonCatalogProvider {
  const service = getServiceInstance();

  // Pure forwarder - no logic
  return {
    getLessonDetail: service.getLessonDetail.bind(service),
    browseLessons: service.browseLessons.bind(service),
    searchLessons: service.searchLessons.bind(service),
    getPopularLessons: service.getPopularLessons.bind(service),
    getCatalogStatistics: service.getCatalogStatistics.bind(service),
    refreshCatalog: service.refreshCatalog.bind(service),
    checkHealth: service.checkHealth.bind(service),
  };
}

// Export types for other modules
export type {
  LessonSummary,
  LessonDetail,
  BrowseLessonsResponse,
  LessonFilters,
  CatalogStatistics,
  PaginationInfo,
} from './models';
